File: LAB_1/sorting.py
import sys
# Append the utils directory to sys.path
sys.path.append(r"D:\program_file\DAA_lab\DAA_files\utils")


import random as r
from utils.timer import Timer  # Assuming you have a Timer class (not included here)
import matplotlib.pyplot as plt


def selection_sort(arr: list[int]) -> list[int]:

    for i in range(len(arr)):
        min_index = i
        for j in range(i + 1, len(arr)):
            if arr[j] < arr[min_index]:
                min_index = j
        arr[i], arr[min_index] = arr[min_index], arr[i]
    return arr


def bubble_sort(arr: list[int]) -> list[int]:
    
    swapped = True 
    while swapped:
        swapped = False
        for i in range(len(arr) - 1):
            if arr[i] > arr[i + 1]:
                arr[i], arr[i + 1] = arr[i + 1], arr[i]
                swapped = True
    return arr


def insertion_sort(arr: list[int]) -> list[int]:

    for i in range(1, len(arr)):
        key = arr[i]
        j = i - 1
        while j >= 0 and key < arr[j]:
            arr[j + 1] = arr[j]
            j -= 1
        arr[j + 1] = key
    return arr


def main():
    test_cases = {}
    for size in [10, 20, 30, 40, 50, 100, 1000, 10000]:
        test_cases[str(size)] = [r.randint(0, 100000) for _ in range(size)]

    x_values = list(test_cases.keys())
    y_values_bubble = []
    y_values_selection = []
    y_values_insertion = []

    print("Bubble Sort:")
    for size, data in test_cases.items():
        timer = Timer()
        timer.start()
        bubble_sort(data.copy())  
        timer.stop()
        print(f"{size}: {timer.elapsed()}")
        y_values_bubble.append(timer.elapsed())

    print("\nSelection Sort:")
    for size, data in test_cases.items():
        timer = Timer()
        timer.start()
        selection_sort(data.copy())  
        timer.stop()
        print(f"{size}: {timer.elapsed()}")
        y_values_selection.append(timer.elapsed())

    print("\nInsertion Sort:")
    for size, data in test_cases.items():
        timer = Timer()
        timer.start()
        insertion_sort(data.copy()) 
        timer.stop()
        print(f"{size}: {timer.elapsed()}")
        y_values_insertion.append(timer.elapsed())

    plt.plot(x_values, y_values_bubble, label="Bubble Sort")
    plt.plot(x_values, y_values_selection, label="Selection Sort")
    plt.plot(x_values, y_values_insertion, label="Insertion Sort")
    plt.xlabel("List Size (n)")
    plt.ylabel("Execution Time (seconds)")
    plt.title("Comparison of Sorting Algorithms")
    plt.legend()
    plt.show()


if __name__ == "__main__":
    main()

File: LAB_1/sorting.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

void selection_sort(int arr[], int n)
{
    int i, j, min_index;
    for (i = 0; i < n - 1; i++)
    {
        min_index = i;
        for (j = i + 1; j < n; j++)
        {
            if (arr[j] < arr[min_index])
            {
                min_index = j;
            }
        }
        if (min_index != i)
        {
            int temp = arr[i];
            arr[i] = arr[min_index];
            arr[min_index] = temp;
        }
    }
}

void bubble_sort(int arr[], int n)
{
    int i, j, swapped;

    for (i = 0; i < n - 1; i++)
    {
        swapped = 0;
        for (j = 0; j < n - i - 1; j++)
        {
            if (arr[j] > arr[j + 1])
            {
                int temp = arr[j];
                arr[j] = arr[j + 1];
                arr[j + 1] = temp;
                swapped = 1;
            }
        }
        if (!swapped)
        {
            break;
        }
    }
}

void insertion_sort(int arr[], int n)
{
    int i, j, key;
    for (i = 1; i < n; i++)
    {
        key = arr[i];
        j = i - 1;
        while (j >= 0 && arr[j] > key)
        {
            arr[j + 1] = arr[j];
            j = j - 1;
        }
        arr[j + 1] = key;
    }
}

// Function to generate random numbers and fill the array
void generate_random_array(int arr[], int n)
{
    for (int i = 0; i < n; i++)
    {
        arr[i] = rand() % 100; 
    }
}
int main()
{
    srand(time(0));
    int sizes[] = {10, 20, 30, 40};

    for (int k = 0; k < 4; k++)
    {
        int n = sizes[k];

        int arr1[n], arr2[n], arr3[n];
        generate_random_array(arr1, n);
        generate_random_array(arr2, n);
        generate_random_array(arr3, n);
        
        /*clock_t start_time = clock(); 
        selection_sort(arr1, n);
        clock_t end_time = clock(); 
        double time_taken = ((double)(end_time - start_time)) / CLOCKS_PER_SEC;
        printf("Selection Sort on array of size %d took %f seconds\n", n, time_taken);*/

        clock_t start_time = clock(); 
        bubble_sort(arr2, n);
        clock_t end_time = clock(); 
        double time_taken = ((double)(end_time - start_time)) / CLOCKS_PER_SEC;
        printf("Bubble Sort on array of size %d took %f seconds\n", n, time_taken);

        /* start_time = clock(); 
        insertion_sort(arr3, n);
        end_time = clock(); 
        time_taken = ((double)(end_time - start_time)) / CLOCKS_PER_SEC;
        printf("Insertion Sort on array of size %d took %f seconds\n", n, time_taken);*/
    }

    return 0;
}
